esphome/components/font/__init__.py

```diff
@@ -228,27 +252,119 @@ def _file_schema(value):
 
 CONFIG_SCHEMA = cv.All(validate_pillow_installed, FONT_SCHEMA)
 
+# PIL doesn't provide a consistent interface for both TrueType and bitmap
+# fonts. So, we use our own wrappers to give us the consistency that we need.
 
-async def to_code(config):
+
+class TrueTypeFontWrapper:
+    def __init__(self, font):
+        self.font = font
+
+    def getoffset(self, glyph):
+        _, (offset_x, offset_y) = self.font.font.getsize(glyph)
+        return offset_x, offset_y
+
+    def getmask(self, glyph, **kwargs):
+        return self.font.getmask(glyph, **kwargs)
+
+    def getmetrics(self, glyphs):
+        return self.font.getmetrics()
+
+
+class BitmapFontWrapper:
+    def __init__(self, font):
+        self.font = font
+        self.max_height = 0
+
+    def getoffset(self, glyph):
+        return 0, 0
+
+    def getmask(self, glyph, **kwargs):
+        return self.font.getmask(glyph, **kwargs)
+
+    def getmetrics(self, glyphs):
+        max_height = 0
+        for glyph in glyphs:
+            mask = self.getmask(glyph, mode="1")
+            _, height = mask.size
+            if height > max_height:
+                max_height = height
+        return (max_height, 0)
+
+
+def convert_bitmap_to_pillow_font(filepath):
+    from PIL import PcfFontFile, BdfFontFile
+
+    local_bitmap_font_file = _compute_local_font_dir(filepath) / os.path.basename(
+        filepath
+    )
+
+    copy_file_if_changed(filepath, local_bitmap_font_file)
+
+    with open(local_bitmap_font_file, "rb") as fp:
+        try:
+            try:
+                p = PcfFontFile.PcfFontFile(fp)
+            except SyntaxError:
+                fp.seek(0)
+                p = BdfFontFile.BdfFontFile(fp)
+
+            # Convert to pillow-formatted fonts, which have a .pil and .pbm extension.
+            p.save(local_bitmap_font_file)
+        except (SyntaxError, OSError) as err:
+            raise cv.Invalid(f"Failed to parse as bitmap font: '{filepath}': {err}")
```

Comment:
```suggestion
            raise EsphomeError(f"Failed to parse as bitmap font: '{filepath}': {err}")
```

`cv.Invalid` is for validation errors, but this runs after the validation phase in `to_code`